### PR TITLE
WDIP-17: Different phase duration for retreats/builds/missed turns

### PR DESCRIPTION
**Pull request by adamlerer**
image

image

image

### CODE REVIEW

**Comment by utiq** [comment]
@adamlerer what else needs to be completed?

**Comment by adamlerer** [comment]
> @adamlerer what else needs to be completed?

This is just a sketch of how you might implement this. The main things that are left are to fix up how the phase length is displayed in all the different places (old UI, new UI, game list, etc.), figure out interactions with the admin control panel actions, and testing that it works (I didn't test this at all).

I found that a good way to find potential places where changes are needed is with the following query: `~/git/webDiplomacy$ grep phaseMinutes *.php */*.php */*/*.php`

File: gamemaster/game.php [hunk 1]
```diff
@@ -548,7 +549,7 @@ private function switchPhaseTime()
 		$this->phaseMinutes = $newPhaseMinutes;
 		$this->processTime = time()+($newPhaseMinutes*60);
 
-		$DB->sql_put("UPDATE wD_Games SET phaseMinutes = ".$this->phaseMinutes.", processTime = ".$this->processTime." WHERE id = ".$this->id);
+		$DB->sql_put("UPDATE wD_Games SET phaseMinutes = ".$this->phaseMinutes.", phaseMinutesRB = -1, processTime = ".$this->processTime." WHERE id = ".$this->id);
```
[the remark this answers is not in the document]
Comment (adamlerer) [reply gamemaster/game.php:552]: Yeah... I was thinking that the admin actions were a less urgent piece and could be fixed up when necessary, and most of them shouldn't need to affect RB. But I'll take another look.

File: locales/English/gamecreate.php [hunk 2]
```diff
@@ -120,7 +120,39 @@
...
+
+			</br></br>
+			<strong>Phase length (Retreats, Builds & Missed Turns)</strong>
+			<img id = "modBtnPhaseLengthRB" height="16" width="16" src="images/icons/help.png" alt="Help" title="Help" />
+			<div id="phaseLengthRBModal" class="modal">
+				<!-- Modal content -->
+				<div class="modal-content">
+					<span class="close9">&times;</span>
+					<p><strong>Phase length (Retreats, Builds & Missed Turns): </strong></br>
+						How long retreat and build phases, and grace periods for missed turns will last.
```
[the remark this answers is not in the document]
Comment (adamlerer) [reply locales/English/gamecreate.php:133]: Missed turns in general.

File: gamecreate.php [hunk 3]
```diff
@@ -98,6 +98,15 @@
 		{
 			throw new Exception(l_t("The phase value is too large or small; it must be between 5 minutes and 10 days."));
 		}
+		$input['phaseMinutesRB'] = (int)$input['phaseMinutesRB'];
+		if ( $input['phaseMinutesRB'] != -1 && ($input['phaseMinutesRB'] < 1 or $input['phaseMinutesRB'] > 1440*10 ))
+		{
+			throw new Exception(l_t("The phase value for retreats and builds is too large or small; it must be between 1 minute and 10 days."));
+		}
+		if ( $input['phaseMinutesRB'] != -1 && $input['phaseMinutesRB'] > $input['phaseMinutes'])
```
[the remark this answers is not in the document]
Comment (adamlerer) [reply gamecreate.php:106]: Hmm, interesting point. Do you think I should have some restriction like `0.1 * phaseMinutes <= phaseMinutesRB <= phaseMinutes`?

I don't think it ever makes sense for phaseMinutesRB to be longer, and I explicitly disallow that here (although you can still enter it in the form :( ).

**Pull request merged**